File: rtix/ipc/tests/test_node.cpp
// Copyright (c) 2025 Velodex Robotics, Inc and RTIX Developers.
// Licensed under Apache-2.0. http://www.apache.org/licenses/LICENSE-2.0

#include <google/protobuf/wrappers.pb.h>
#include <gtest/gtest.h>
#include <thread>
#include "rtix/core/timer.h"
#include "rtix/ipc/node.h"
#include "rtix/ipc/tests/test_channel_map.h"

using google::protobuf::Int64Value;
using rtix::core::Timer;
using rtix::ipc::ChannelMap;
using rtix::ipc::Node;

void node_a_callback(ChannelMap channel_map) {
  EXPECT_TRUE(channel_map.nodes.find("node_a") != channel_map.nodes.end());
  auto node_a = Node(channel_map.getNodeConfig("node_a"));

  // Access the message publisher
  auto node_a_int = node_a.publisher("node_a_int");

  // Sleep before sending to give the receiver time to spin up and listen
  Timer::Sleep(0.5);

  Int64Value pb_int{};
  pb_int.set_value(11);
  node_a_int->send(pb_int);
  Timer::Sleep(0.5);
  node_a_int->send(pb_int);
}

void node_b_callback(ChannelMap channel_map) {
  EXPECT_TRUE(channel_map.nodes.find("node_b") != channel_map.nodes.end());
  auto node_b = Node(channel_map.getNodeConfig("node_b"));

  // Access the message subscriber
  auto node_a_int = node_b.subscriber("node_a_int");

  // Listen for the message and check the correct value was received
  Int64Value pb_int{};
  EXPECT_TRUE(node_a_int->recv(pb_int));
  EXPECT_EQ(pb_int.value(), 11);
}

TEST(Ipc, Node) {
  auto channel_map = load_test_channel_map();

  std::thread node_a_thread(node_a_callback, channel_map);
  std::thread node_b_thread(node_b_callback, channel_map);

  node_a_thread.join();
  node_b_thread.join();
}

File: rtix/ipc/tests/test_node.py
# Copyright (c) 2025 Velodex Robotics, Inc and RTIX Developers.
# Licensed under Apache-2.0. http://www.apache.org/licenses/LICENSE-2.0

import threading
import time
from google.protobuf.wrappers_pb2 import Int64Value

from rtix.ipc.node import Node
from rtix.ipc.channel_map import ChannelMap
from rtix.ipc.tests.test_channel_map import load_test_channel_map


class ThreadWithAssertion(threading.Thread):

    def __init__(self, target, channel_map):
        super().__init__(target=target, args=(channel_map, ))
        self.exc = None

    def run(self):
        try:
            super().run()
        except BaseException as e:
            self.exc = e

    def join(self, timeout=None):
        super().join(timeout)
        if self.exc:
            raise self.exc


def test_node():
    # Assertions within a thread won't automatically fail the pytest test
    # because they are raised in a different context. To correctly capture and
    # propagate these failures, a mechanism is needed to re-raise exceptions in
    # the main thread.  We subclass Thread to raise an exception on join, and
    # pass the thread callbacks in the main loop below.

    def node_a_callback(channel_map: ChannelMap):
        assert "node_a" in channel_map.nodes
        node_a = Node(config=channel_map.nodes["node_a"])

        # Access the message publisher
        node_a_int = node_a.publisher(channel_id="node_a_int")

        # Sleep before sending to give the receiver time to spin up and listen
        time.sleep(0.1)
        node_a_int.send(msg=Int64Value(value=11))

    def node_b_callback(channel_map: ChannelMap):
        assert "node_b" in channel_map.nodes
        node_b = Node(config=channel_map.nodes["node_b"])

        # Access the message subscriber
        node_a_int = node_b.subscriber(channel_id="node_a_int")

        # Listen for the message and check the correct value was received
        pb_int = Int64Value()
        assert node_a_int.recv(msg=pb_int) == True
        assert pb_int.value == 11

    channel_map = load_test_channel_map()

    node_a_thread = ThreadWithAssertion(
        target=node_a_callback,
        channel_map=channel_map,
    )
    node_b_thread = ThreadWithAssertion(
        target=node_b_callback,
        channel_map=channel_map,
    )

    node_a_thread.start()
    node_b_thread.start()

    node_a_thread.join()
    node_b_thread.join()
